community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ImportLogic.java

```diff
@@ -0,0 +1,528 @@
+                    currentSetOfRelationshipsMemoryUsage > 0 )
+            {
+                // OK the current set of types is enough to fill the cache
+                break;
+            }
+
+            currentSetOfRelationshipsMemoryUsage += memoryUsageForThisType;
+        }
+        return toType;
+    }
+
+    /**
+     * Optimizes the relationship groups store by physically locating groups for each node together.
+     */
+    public void defragmentRelationshipGroups()
+    {
+        // Defragment relationships groups for better performance
+        new RelationshipGroupDefragmenter( config, executionMonitor, numberArrayFactory )
+                .run( max( maxMemory, peakMemoryUsage ), neoStore, neoStore.getNodeStore().getHighId() );
+    }
+
+    /**
+     * Builds the counts store. Requires that {@link #importNodes()} and {@link #importRelationships()} has run.
+     */
+    public void buildCountsStore()
+    {
+        // Count nodes per label and labels per node
+        try ( CountsAccessor.Updater countsUpdater = neoStore.getCountsStore().reset(
+                neoStore.getLastCommittedTransactionId() ) )
+        {
+            MigrationProgressMonitor progressMonitor = new SilentMigrationProgressMonitor();
+            nodeLabelsCache = new NodeLabelsCache( numberArrayFactory, neoStore.getLabelRepository().getHighId() );
+            memoryUsageStats = new MemoryUsageStatsProvider( nodeLabelsCache );
+            executeStage( new NodeCountsStage( config, nodeLabelsCache, neoStore.getNodeStore(),
+                    neoStore.getLabelRepository().getHighId(), countsUpdater, progressMonitor.startSection( "Nodes" ),
+                    memoryUsageStats ) );
+            // Count label-[type]->label
+            executeStage( new RelationshipCountsStage( config, nodeLabelsCache, neoStore.getRelationshipStore(),
+                    neoStore.getLabelRepository().getHighId(),
+                    neoStore.getRelationshipTypeRepository().getHighId(),
+                    countsUpdater, numberArrayFactory, progressMonitor.startSection( "Relationships" ) ) );
+        }
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        // We're done, do some final logging about it
+        long totalTimeMillis = currentTimeMillis() - startTime;
+        executionMonitor.done( totalTimeMillis, format( "%n%s%nPeak memory usage: %s", storeUpdateMonitor, bytes( peakMemoryUsage ) ) );
+        log.info( "Import completed successfully, took " + duration( totalTimeMillis ) + ". " + storeUpdateMonitor );
+
+        if ( nodeRelationshipCache != null )
+        {
+            nodeRelationshipCache.close();
+        }
+        if ( nodeLabelsCache != null )
+        {
+            nodeLabelsCache.close();
+        }
+        if ( idMapper != null )
+        {
+            idMapper.close();
+        }
+        inputCache.close();
+        // TODO close badCollector here instead of in import tool?
```

Comment:
TODO

Reply:
Some idea about cleaning up relationship between some classes, there's a card already for it tho.